addons/stock/models/stock_move_line.py

```diff
@@ -430,14 +430,17 @@ def _action_done(self):
                             # If a picking type is linked, we may have to create a production lot on
                             # the fly before assigning it to the move line if the user checked both
                             # `use_create_lots` and `use_existing_lots`.
-                            if ml.lot_name and not ml.lot_id:
-                                lot = self.env['stock.production.lot'].search([
-                                    ('company_id', '=', ml.company_id.id),
-                                    ('product_id', '=', ml.product_id.id),
-                                    ('name', '=', ml.lot_name),
-                                ])
-                                if lot:
-                                    ml.lot_id = lot.id
+                            if ml.lot_name:
+                                if ml.product_id.tracking == 'lot' and not ml.lot_id:
+                                    lot = self.env['stock.production.lot'].search([
+                                        ('company_id', '=', ml.company_id.id),
+                                        ('product_id', '=', ml.product_id.id),
+                                        ('name', '=', ml.lot_name),
+                                    ])
```

Comment:
use `limit=1` for better performance

Reply:
It also makes it safer in case duplicates exist! Added it in, thanks.